apps/prairielearn/src/sync/sharing.ts

```diff
@@ -185,3 +185,45 @@ export async function checkInvalidSharingSetRemovals(
 
   return existInvalidSharingSetRemovals;
 }
+
+export function checkInvalidSharedAssessments(
+  courseData: CourseData,
+  logger: ServerJobLogger,
+): boolean {
+  const invalidSharedAssessments: string[] = [];
+  for (const courseInstanceKey in courseData.courseInstances) {
+    const courseInstance = courseData.courseInstances[courseInstanceKey];
+    for (const tid in courseInstance.assessments) {
+      const assessment = courseInstance.assessments[tid];
+      if (!assessment?.data?.shareSourcePublicly) {
+        continue;
+      }
+      if (!(assessment.data && assessment.data.zones)) {
+        continue;
+      }
+      for (const zone of assessment.data.zones) {
+        if (!zone.questions) {
+          continue;
+        }
+        for (const question of zone.questions) {
```

Comment:
this code could be made a bit more compact. For example, if you change the for loop header to 

```js
for (const question of (zone?.questions ?? []))
```
then you can just leave out the if statement entirely. There are a couple of other places in this function where you could use tricks like this to make the codes shorter.

Reply:
```suggestion
for (const question of (zone?.questions ?? []))
```

Reply:
Changed